src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -1841,114 +1841,153 @@ void TransferTextSpanPosToRunTextPos()
 
-                // Label to jump to when any branch completes successfully.
-                Label doneAlternate = DefineLabel();
-
-                // A failure in a branch other than the last should jump to the next
-                // branch, not to the final done.
-                Label originalDoneLabel = doneLabel;
+                // After executing the alternation, subsequent matching may fail, at which point execution
+                // will need to backtrack to the alternation.  We emit a branching table at the end of the
+                // alternation, with a label that will be left as the "doneLabel" upon exiting emitting the
+                // alternation.  The branch table is populated with an entry for each branch of the alternation,
+                // containing either the label for the last backtracking construct in the branch if such a construct
+                // existed (in which case the doneLabel upon emitting that node will be different from before it)
+                // or the label for the next branch.
+                var labelMap = new Label[childCount];
+                Label backtrackLabel = DefineLabel();
 
-                int childCount = node.ChildCount();
-                for (int i = 0; i < childCount - 1; i++)
+                for (int i = 0; i < childCount; i++)
                 {
-                    Label nextBranch = DefineLabel();
-                    doneLabel = nextBranch;
+                    bool isLastBranch = i == childCount - 1;
+
+                    Label nextBranch = default;
+                    if (!isLastBranch)
+                    {
+                        // Failure to match any branch other than the last one should result
+                        // in jumping to process the next branch.
+                        nextBranch = DefineLabel();
+                        doneLabel = nextBranch;
+                    }
+                    else
+                    {
+                        // Failure to match the last branch is equivalent to failing to match
+                        // the whole alternation, which means those failures should jump to
+                        // what "doneLabel" was defined as when starting the alternation.
+                        doneLabel = originalDoneLabel;
+                    }
 
                     // Emit the code for each branch.
                     EmitNode(node.Child(i));
 
+                    // Add this branch to the backtracking table.  At this point, either the child
+                    // had backtracking constructs, in which case doneLabel points to the last one
+                    // and that's where we'll want to jump to, or it doesn't, in which case doneLabel
+                    // still points to the nextBranch, which similarly is where we'll want to jump to.
+                    if (!isAtomic)
+                    {
+                        EmitRunstackResizeIfNeeded(3);
+                        EmitRunstackPush(() => Ldc(i));
```

Comment:
Nit: no need to change this now, but at least when doing the original Span changes it helped a lot having comments explaining what this would do like:
```c#
base.runstack[runstackpos++] = 0;
base.runstack[runstackpos++] = alternation_starting_crawlpos;
base.runstack[runstackpos++] = alternation_starting_runtextpos;
```

I can come back later and add these for all these places that have them missing.

Reply:
Thanks. I'll add a bunch of comments in my next PR.